runtime/ops/net.rs

```diff
@@ -528,3 +546,302 @@ fn op_listen(
+      }
+    }
+  }
+
+  #[derive(Deserialize)]
+  #[serde(rename_all = "camelCase")]
+  struct NameServer {
+    ip_addr: String,
+    #[serde(default = "default_port")]
+    port: u16,
+    #[serde(default)]
+    protocol: Protocol,
+  }
+
+  let ResolveAddrArgs {
+    query,
+    record_type,
+    options,
+  } = serde_json::from_value(args)?;
+
+  // Check for permission. If name server option is provided, use them for checking.
+  // Otherwise, checks if `allow-net` is present without any specific allowlist.
+  {
+    let perm_args = options
+      .as_ref()
+      .and_then(|opt| opt.name_server.as_ref())
+      .map(|ns| (&ns.ip_addr, ns.port));
+
+    let s = state.borrow();
+    let perm = s.borrow::<Permissions>();
+
+    if let Some((ip_addr, port)) = perm_args {
+      perm.check_net(ip_addr, port)?;
+    } else {
+      perm.check_net_all()?;
+    }
+  }
+
+  let resolver = if let Some(options) = options {
+    let conf = if let Some(name_server) = options.name_server {
+      let mut c = ResolverConfig::new();
+      c.add_name_server(NameServerConfig {
+        socket_addr: resolve_addr(&name_server.ip_addr, name_server.port)
+          .await?
+          .next()
+          .ok_or_else(|| generic_error("Invalid IP address or port"))?,
+        protocol: name_server.protocol.into(),
+        tls_dns_name: None,
+      });
+      c
+    } else {
+      let (c, _) = read_system_conf()?;
+      c
+    };
+    AsyncResolver::tokio(conf, ResolverOpts::default()).await?
+  } else {
+    AsyncResolver::tokio_from_system_conf().await?
+  };
+
+  let results: Vec<DnsReturnRecord> = resolver
+    .lookup(query, record_type.into(), Default::default())
+    .await?
+    .iter()
+    .filter_map(rdata_to_return_record(record_type))
+    .collect();
+
+  Ok(json!(results))
+}
+
+fn rdata_to_return_record(
+  ty: SupportedRecordType,
+) -> impl Fn(&RData) -> Option<DnsReturnRecord> {
+  use SupportedRecordType::*;
+  move |r: &RData| -> Option<DnsReturnRecord> {
+    match ty {
+      A => r.as_a().map(ToString::to_string).map(DnsReturnRecord::A),
+      AAAA => r
+        .as_aaaa()
+        .map(ToString::to_string)
+        .map(DnsReturnRecord::AAAA),
+      ANAME => r
+        .as_aname()
+        .map(ToString::to_string)
+        .map(DnsReturnRecord::ANAME),
+      CNAME => r
+        .as_cname()
+        .map(ToString::to_string)
+        .map(DnsReturnRecord::CNAME),
+      MX => r.as_mx().map(|mx| DnsReturnRecord::MX {
+        preference: mx.preference(),
+        exchange: mx.exchange().to_string(),
+      }),
+      PTR => r
+        .as_ptr()
+        .map(ToString::to_string)
+        .map(DnsReturnRecord::PTR),
+      SRV => r.as_srv().map(|srv| DnsReturnRecord::SRV {
+        priority: srv.priority(),
+        weight: srv.weight(),
+        port: srv.port(),
+        target: srv.target().to_string(),
+      }),
+      TXT => r.as_txt().map(|txt| {
+        // TODO(magurotuna): parses bytes as UTF-8, is it alright?
```

Comment:
Good question. I think the RFC suggests that octets should be in the range 0x20..=0x7E but the real world is probably more... incalculable. I'd play it safe and interpret it as Latin-1, example here:
https://github.com/denoland/deno/blob/b1230a85e8ba63f3542892711043cefdc916acf8/op_crates/fetch/lib.rs#L173-L177

Reply:
I see, Latin-1 seems reasonable.

Reply:
Done in 20a42de